scRNA plant tutorial (#2435)
galaxyproject/training-material: topics/transcriptomics/tutorials/scrna-plant/tutorial.md

>    - {% icon param-file %} *"Annotated data matrix"*: `anndata` (output of **Manipulate AnnData** {% icon tool %})
>    - *"Method used for plotting"*: `Generic: Violin plot, using 'pl.violin'`
>      - *"Keys for accessing variables"*: `Subset of variables in 'adata.var_names' or fields of '.obs'`
>        - *"Keys for accessing variables"*: `n_genes_by_counts, total_counts`
>      - *"The key of the observation grouping to consider"*: `batch`
>      - In *"Violin plot attributes"*:
>        - *"Add a stripplot on top of the violin plot"*: `Yes`
>          - *"Add a jitter to the stripplot"*: `Yes`
>            - *"Size of the jitter points"*: `0.4`
>        - *"Display keys in multiple panels"*: `Yes`
>
{: .hands_on}

This should result in the following violin plot, where we see the distribution of the number of genes per cell (`n_genes_by_counts`) and the library size per cell, where the library size is the total number of mRNA in a cell, regardless of which gene it came from (`total_counts`).

!["Violin Plots per Batch"](../../images/scrna-plant/violin1.png)

> Each dot is a cell, and the x-axis has two groups showing the absolute values of for each batch (`shr` and `wt`). The y-values correspond to the actual value of each cell, but the x-vales within each group are just random jitter to help see the cells more clearly. The outline of the violin plot shows us the density of each batch, where we can see that most of the cells in the `shr` group have an average number of features just below 2000 genes, compared to `wt` which has an average number of features just below 4000 genes. If we compare the library size however, we notice that the outlines are very similar and that the average library size across batches are comparable.

Reviewer: ```suggestion
Each dot is a cell, and the x-axis has two groups showing the absolute values of for each batch (`shr` and `wt`). The y-values correspond to the actual value of each cell, but the x-vales within each group are just random jitter to help see the cells more clearly. The outline of the violin plot shows us the density of each batch, where we can see that most of the cells in the `shr` group have an average number of features just below 2000 genes, compared to `wt` which has an average number of features just below 4000 genes. If we compare the library size, however, we notice that the outlines are very similar and that the average library sizes across batches are comparable.
```

### Filter

Most scRNA-seq datasets need to set a minimum threshold of detectability in order to ensure that the cells are of high enough quality to be used for analysis. Conversely, sometimes an upper threshold also should be set to counter the case where a cell is a cell-doublet (in the sense that two cells were captured and sequenced accidentally as a single cell). 

> The above violin plots suggest quite a few outliers, mostly on the upper-end of the density plots, suggesting a few cell-doublets in the data. The effect of these is not something we cannot easily regress out, and so we must filter them out instead. By looking at the plots, we can see that a good maximum library size here would be 100 000 transcripts, and a good maximum feature size would be 10 000 transcripts.

Reviewer: ```suggestion
The above violin plots suggest quite a few outliers, mostly on the upper-end of the density plots, suggesting a few cell-doublets in the data. The effect of these is not something we cannot easily regress out, and so we must filter them out instead. By looking at the plots, we can see that a good maximum library size here would be 100 000 transcripts, and a good maximum feature size would be 10000 transcripts.
```

Author: So I tend to leave a space, because my eyes suffer otherwise and in the previous part I say 100 000. I think I want to keep this

> For this analysis we will set a minimum threshold of detectability that each cell should count transcripts from at least 200 genes, and that each gene should be expressed in at least 5 cells. These lower-bound thresholds are not so easy to derive from such QC plots, but depend mostly on the type of data you are using. If the number of features and the library sizes were one order of magnitude higher, one might consider also scaling the lower-bound thresholds by one order of magnitude.

Reviewer: ```suggestion
For this analysis, we will set a minimum threshold of detectability that each cell should count transcripts from at least 200 genes, and that each gene should be expressed in at least 5 cells. These lower-bound thresholds are not so easy to derive from such QC plots, but depend mostly on the type of data you are using. If the number of features and the library sizes were one order of magnitude higher, one might consider also scaling the lower-bound thresholds by one order of magnitude.
```

> ### {% icon hands_on %} Hands-on: Filtering
>
> 1. {% tool [Filter](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_filter/scanpy_filter/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Inspect and manipulate** {% icon tool %})
>    - *"Method used for filtering"*: `Filter cell outliers based on counts and numbers of genes expressed, using 'pp.filter_cells'`
>        - *"Filter"*: `Minimum number of genes expressed`
>            - *"Minimum number of genes expressed required for a cell to pass filtering"*: `200`
>

> > 1. {% tool [Filter](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_filter/scanpy_filter/1.7.1+galaxy0) %} with the following parameters:

Reviewer: the numbering is 1 in all the step of this hands-on

Author: Oh whoops, I left a new line too many -- thanks

>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Filter** {% icon tool %})
>    - *"Method used for filtering"*: `Filter cell outliers based on counts and numbers of genes expressed, using 'pp.filter_cells'`
>        - *"Filter"*: `Maximum number of counts`
>            - *"Maximum number of counts required for a cell to pass filtering"*: `100000`
>
{: .hands_on}

If we inspect the resulting dataset, we see that the number of cells remaining is `5826` and the number of genes is `18913`, which is still a good number of cells and genes. It is good practice to save the raw data in case we need to use it data, so we "freeze the raw data" into a slot for later safe keeping.

> ### {% icon hands_on %} Hands-on: Saving the original matrix
>
> 1. {% tool [Manipulate AnnData](toolshed.g2.bx.psu.edu/repos/iuc/anndata_manipulate/anndata_manipulate/0.7.5+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Filter** {% icon tool %})
>    - *"Function to manipulate the object"*: `Freeze the current state into the 'raw' attribute`
>
{: .hands_on}

## Confounder Removal

> Now we have a set of cells we are reasonably confident contain desirable biological variation (please see the ["An introduction to scRNA-seq data"]({% link videos/#video-transcriptomics-scrna-intro %}) at time point 11:28 as a reference for the types of wanted and unwanted variation). We wish to now regress out the library size variation, as it is not indicative of cell type. To do this we will first ensure that all cells are of the same comparable library size by applying individual size factors to each cell. 

Reviewer: ```suggestion
Now we have a set of cells we are reasonably confident contain desirable biological variation (please see the ["An introduction to scRNA-seq data"]({% link videos/#video-transcriptomics-scrna-intro %}) at time point 11:28 as a reference for the types of wanted and unwanted variation). We wish to now regress out the library size variation, as it is not indicative of cell type. To do this, we will first ensure that all cells are of the same comparable library size by applying individual size factors to each cell. 
```

>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Normalize** {% icon tool %})
>    - *"Method used for inspecting"*: `Logarithmize the data matrix, using 'pp.log1p'`
>
> 1. {% tool [Remove confounders](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_remove_confounders/scanpy_remove_confounders/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Inspect and manipulate** {% icon tool %})
>    - *"Method used for plotting"*: `Regress out unwanted sources of variation, using 'pp.regress_out'`
>        - *"Keys for observation annotation on which to regress on"*: `total_counts`
>
> 1. {% tool [Inspect and manipulate](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_inspect/scanpy_inspect/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Remove confounders** {% icon tool %})
>    - *"Method used for inspecting"*: `Scale data to unit variance and zero mean, using 'pp.scale'`
>        - *"Maximum value"*: `10.0`
>
{: .hands_on}

> ## Dimension Reduction

Reviewer: dimensionality?

Author: I think Dimension is okay

Reviewer 2: both ok, but I think consistency is key here. You've used dimensionality above in some places where if you use 'dimension reduction' here, you could be using dimension above. Maybe consider trying to unify those?

> Dimension reduction is the art of reducing a high dimensional dataset into a low dimensional "embedding" that humans can actually see (i.e. 2 or 3 dimensions), ideally such that the relationships or distances between data points are preserved in this embedding. In the context of single-cell datasets, this essentially means compressing > 10 000 genes into just 2 X/Y variables. 

Reviewer: ```suggestion
Dimension reduction is the art of reducing a high dimensional dataset into a low dimensional "embedding" that humans can actually see (i.e. 2 or 3 dimensions), ideally such that the relationships or distances between data points are preserved in this embedding. In the context of single-cell datasets, this essentially means compressing > 10000 genes into just 2 X/Y variables. 
```

> ### {% icon tip %} Tip: Dimension Reduction
>
> You can learn more about dimension reduction by consulting the [An introduction to scRNA-seq data analysis]({% link videos/#video-transcriptomics-scrna-intro %}) at time point 13:46.
>    {: .tip}

This is usually a two step process: 

>  1. A Principal Component Analysis (PCA) is used to perform an optimized "rotation" of the ~ 20 000 "unit" gene axes that we have, in order to better fit the data. These new axes or "principal components" are then *linear* combinations of the original unit axes, with an assosciated score of how variable each new axis is. By sorting these principal components by most variable to least variable, we can select the top N components and discard the rest, leaving us with most of the variation still in the data. Usually we select the top 40 principal components, which from 20 000 is a *huge* reduction in the dimensionality of the dataset. 

Reviewer: ```suggestion
 1. A Principal Component Analysis (PCA) is used to perform an optimized "rotation" of the ~ 20 000 "unit" gene axes that we have, in order to better fit the data. These new axes or "principal components" are then *linear* combinations of the original unit axes, with an associated score of how variable each new axis is. By sorting these principal components by most variable to least variable, we can select the top N components and discard the rest, leaving us with most of the variation still in the data. Usually, we select the top 40 principal components, which from 20000 is a *huge* reduction in the dimensionality of the dataset. 
```

Author: I will fix the associated

> 1. We then perform a more complex kind of dimension reduction, one that does not assume any linearality in the axes. For scRNA-seq this is either tSNE or UMAP, with UMAP being the main choice due to how flexible it is in incorporating new data. Though UMAP is capable of working on extremely high dimensional datasets, it is often limited by time and space constraints (read: the computer does not respond in a reasonable timeframe, or it crashes), and so therefore it is quite normal to feed UMAP the output of the PCA as input. This will then be a dimension reduction from 50 to 2. With these final 2 dimensions, we can plot the data.

Reviewer: ```suggestion
1. We then perform a more complex kind of dimension reduction, one that does not assume any linearity in the axes. For scRNA-seq, this is either tSNE or UMAP, with UMAP being the main choice due to how flexible it is in incorporating new data. Though UMAP is capable of working on extremely high dimensional datasets, it is often limited by time and space constraints (read: the computer does not respond in a reasonable timeframe, or it crashes), and so therefore it is quite normal to feed UMAP the output of the PCA as input. This will then be a dimension reduction from 50 to 2. With these final 2 dimensions, we can plot the data.
```

> ### {% icon hands_on %} Hands-on: PCA and UMAP
>

> > 1. {% tool [Cluster, infer trajectories and embed](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_cluster_reduce_dimension/scanpy_cluster_reduce_dimension/1.7.1+galaxy0) %} with the following parameters:

Reviewer: again the numbering is 1 in every step

>    - *"Method used for plotting"*: `Embeddings: Scatter plot in PCA coordinates, using 'pl.pca`
>      - *"Keys for annotations of observations/cells or variables/genes"*: `batch`
>      - In *"Plot attributes"*
>        - *"Colors to use for plotting categorical annotation groups"*: `rainbow`
>
> 1. {% tool [Plot with scanpy](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_plot/scanpy_plot/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Cluster, infer trajectories and embed** {% icon tool %})
>    - *"Method used for plotting"*: `Embeddings: Scatter plot in UMAP basis, using 'pl.umap'`
>      - *"Keys for annotations of observations/cells or variables/genes"*: `batch`
>      - In *"Plot attributes"*
>        - *"Colors to use for plotting categorical annotation groups"*: `rainbow`
>
{: .hands_on}

| PCA                                | UMAP                                 |
|:----------------------------------:|:------------------------------------:|
| ![PCA](../../images/pca_batch.png) | ![UMAP](../../images/umap_batch.png) |

> From this we can see that there is a reasonable overlap in our batches shown both in the PCA and UMAP embeddings. This is good because it shows that though there is *some* batch effect (i.e. cells from one batch appear to cluster on a different side of the plot than the other) it is not significant enough for there not to be some commonality between the batches. 

Reviewer: ```suggestion
From this, we can see that there is a reasonable overlap in our batches shown both in the PCA and UMAP embeddings. This is good because it shows that though there is *some* batch effect (i.e. cells from one batch appear to cluster on a different side of the plot than the other) it is not significant enough for there not to be some commonality between the batches. 
```

There are batch correction algorithms for cases where one batch clusters completely seperate from the other, but this is not necessary here. 

## Clustering

Let us cluster the cells and see what cell types we can discover in the plots. The cell types that we wish to recover are those shown in the original paper, as highlighted by the gene markers in the DotPlot below.

| UMAP                                   | DotPlot                                    |
|:--------------------------------------:|:------------------------------------------:|
| ![PCA](../../images/clusters_expected) | ![UMAP](../../images/dotplot_expected.png) |

> > ### {% icon hands_on %} Hands-on: Task description

Reviewer: missing task description

>    - *"Method used for plotting"*: `Embeddings: Scatter plot in UMAP basis, using 'pl.umap'`
>        - *"Keys for annotations of observations/cells or variables/genes"*: `leiden, batch`
>        - *"Show edges?"*: `No`
>        - In *"Plot attributes"*:
>            - *"Location of legend"*: `on data`
>            - *"Legend font size"*: `14`
>            - *"Colors to use for plotting categorical annotation groups"*: `rainbow (Miscellaneous)`
>
>    > ### {% icon comment %} Comment
>    >
>    > We print the legend on the data because it's easier to see where the cluster labels apply.
>    {: .comment} 
>
{: .hands_on}

![UMAP Leiden Clusters](../../images/scrna-plant/clusters_found.png)

Here we have recovered 13 clusters but we don't yet know what types they describe. If we have a list of genes that we know are indicative of a certain cell type (i.e. are marker genes) then we can use this to assign labels to our clusters.

Let us here try to recreate the DotPlot from the paper using the clusters we have discovered.

## DotPlot and Validating Cell Types

> > ### {% icon hands_on %} Hands-on: Task description

Reviewer: missing task description

>            - {% icon param-repeat %} *"Insert Group of variables to highlight"*
>                - *"Start"*: `35`
>                - *"End"*: `39`
>                - *"Label"*: `Xylem`
>            - {% icon param-repeat %} *"Insert Group of variables to highlight"*
>                - *"Start"*: `40`
>                - *"End"*: `41`
>                - *"Label"*: `VC`
>        - *"Custom figure size"*: `No: the figure width is set based on the number of variable names and the height is set to 10.`
>
{: .hands_on}

By running the above we end up with the following DotPlot:

![Dotplot In Analyssis](../../images/dotplot_found.png)

Notice how we have the Columella, QC and NC sharing the same cluster 11, that the Endodermis is localised in cluster 8, the Cortex in cluster 9, the Trichoblast cells in cluster 3, the Xylem showing strong expression in cluster 12, VC in clusters 4 and 6, and the Atrichoblasts have a mixed expression in a range of clusters. These characteristics are the same as the Dotplot in the original paper.

We can use this Dotplot as a guide to relabel our clusters and give more meaningful annotations to the plots.

## Relabel clusters

> > ### {% icon hands_on %} Hands-on: Task description

Reviewer: missing task description

>
> 1. {% tool [Plot with scanpy](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_plot/scanpy_plot/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata` (output of **Manipulate AnnData** {% icon tool %})
>    - *"Method used for plotting"*: `Embeddings: Scatter plot in UMAP basis, using 'pl.umap'`
>        - *"Keys for annotations of observations/cells or variables/genes"*: `leiden, batch`
>        - *"Show edges?"*: `No`
>        - In *"Plot attributes"*:
>            - *"Location of legend"*: `on data`
>            - *"Legend font size"*: `14`
>            - *"Colors to use for plotting categorical annotation groups"*: `rainbow (Miscellaneous)`
>
{: .hands_on}

![Relabelled clusters](../../images/scrna-plant/clusters_found_relabelled.png)

If we look at the clusters now, and compare them to the original image in the paper, we can infer that the meristematic cells are likely to be derived from cluster 1, which shares soft clustering with the trichoblasts suggesting a trajectory pathway which could be explored.

## Conclusions

> In this tutorial we have recapitulated the same clustering analysis in the ["Spatiotemporal Developmental Trajectories in the Arabidopsis Root Revealed Using High-Throughput Single-Cell RNA Sequencing"](https://doi.org/10.1016/j.devcel.2019.02.022) paper, and validated them by comparing DotPlots for specific genes that were used as markers in that paper.

Reviewer: ```suggestion
In this tutorial, we have recapitulated the same clustering analysis in the ["Spatiotemporal Developmental Trajectories in the Arabidopsis Root Revealed Using High-Throughput Single-Cell RNA Sequencing"](https://doi.org/10.1016/j.devcel.2019.02.022) paper, and validated them by comparing DotPlots for specific genes that were used as markers in that paper.
```

> From this point we can perform a lineage analysis to infer a differentiation pathway between the clusters. For ScanPy there is the PAGA option, however this does not work so well with the current dataset, so it is encouraged that users use the original Seurat trajectory suite that was given in the paper, or to experiment with Monocle.

Reviewer: ```suggestion
From this point, we can perform a lineage analysis to infer a differentiation pathway between the clusters. For ScanPy there is the PAGA option, however, this does not work so well with the current dataset, so it is encouraged that users use the original Seurat trajectory suite that was given in the paper, or to experiment with Monocle.
```

Both libraries are available within the RStudio and Jupyter Notebook libraries in the interactive Galaxy envronments which can be found in the *"Miscellaneous Tools"* section under the *"Interactive Tools"* subheading.

This entire tutorial can be invoked from the scRNA Plant Workflow shown below:

![scRNA Plant](../../images/scrna-plant/scrna-plant-workflow.png)

> The input parameters take two datasets as input, and 5 parameters using the defaults shown in this tutorial: min cells, min genes, max features, max library size, and leiden clustering resolution. Play with these options to see how robust the analysis really is!

Reviewer: ```suggestion
The input parameters take two datasets as input, and 5 parameters using the defaults shown in this tutorial: min cells, min genes, max features, max library size, and Leiden clustering resolution. Play with these options to see how robust the analysis really is!
```

---
layout: tutorial_hands_on
title: "Downstream Single-cell RNA Plant analysis with ScanPy"
subtopic: single-cell
priority:

zenodo_link: 'https://zenodo.org/record/4597857'
tags:

>   - single-cell

Reviewer 2: plants?, paper-replication

questions:
  - Can we reclaim cell markers using a different analysis method?
  - Are highly variable genes paramount to the analysis?
objectives:
  - Perform filtering, dimension reduction, and clustering
  - Generate a DotPlot emulating the original paper using a different analysis tool
  - Determine robust clusters across scRNA-seq pipelines
requirements:
  -
    type: "internal"
    topic_name: transcriptomics
    tutorials:
        - scrna-scanpy-pbmc3k

time_estimation: 2H
key_points:
  - Filtering parameters are dependent on the dataset, and should be explored using scatter or violin plots
  - A DotPlot is a fantastic way to validate clusters across different analyses
contributors:
  - mtekman
  - beatrizserrano

gitter: Galaxy-Training-Network/galaxy-single-cell

---

# Introduction

> {:.no_toc}

Reviewer 2: I think I'd just remove it, it's obvious it's intro from being the first paragraph

> ### {% icon comment %} Comment
>
> Please familiarise yourself with the ["Clustering 3K PBMCs with ScanPy"]({% link topics/transcriptomics/tutorials/scrna-scanpy-pbmc3k/tutorial.md %}) tutorial first, as much of the process is the same, and the accompanying slide deck better explains some of the methods and concepts better.
>

> {: .comment}

Reviewer 2: I think this might fit better next to `# Data`, either before or after. No sense warning them before they've read the summary and know what it's about.

Single cell RNA-seq analysis is a cornerstone of developmental research and provides a great level of detail in understanding the underlying dynamic processes within tissues. In the context of plants, this highlights some of the key differentiation pathways that root cells undergo.

![The various cell subpopulations that are expected within this tutorial]({% link topics/transcriptomics/images/scrna-plant/scrna-plant1.jpg %})

> This tutorial replicates the paper ["Spatiotemporal Developmental Trajectories in the Arabidopsis Root Revealed Using High-Throughput Single-Cell RNA Sequencing"](https://doi.org/10.1016/j.devcel.2019.02.022) {% cite denyer2019spatiotemporal %}, where the major plant cell types are recovered in the data as well as distinguishing between QC and meristematic cells. The original paper used the Seurat analysis suite {% cite satija2015spatial %}, but here we will use the ScanPy analysis suite {% cite wolf2018scanpy %} integrated within the single-cell resources in Galaxy {% cite tekman2020single %} .

Reviewer 2: just a stylistic suggestion. Usually we put it in ()s if it just tangentially mentions a citation to discuss something else, and without () if we're talking about "The work Alice did in {% cite %} proves whatever"

```suggestion
This tutorial replicates the paper ["Spatiotemporal Developmental Trajectories in the Arabidopsis Root Revealed Using High-Throughput Single-Cell RNA Sequencing"](https://doi.org/10.1016/j.devcel.2019.02.022) ({% cite denyer2019spatiotemporal %}), where the major plant cell types are recovered in the data as well as distinguishing between QC and meristematic cells. The original paper used the Seurat analysis suite ({% cite satija2015spatial %}), but here we will use the ScanPy analysis suite ({% cite wolf2018scanpy %}) integrated within the single-cell resources in Galaxy ({% cite tekman2020single %}).
```

> > 1. Columns, "ATXG" are the gene prefixes commonly used by the Arabidopsis genomes.
> > 1. There are 1 099 lines and therefore 1 099 cells
> > 1. If we scroll the preview window all the way to the right we see 27 630 as the last number. Given that the first column are the cell names, we must subtract by 1 to get 27 629 cells.
> >
> {: .solution}
>
{: .question}

If the above feels like a convoluted way to get the dimensionality, that's because we haven't imported the data into the right format. For this we need to import both datasets into a single AnnData object (see the [AnnData section in the previous tutorial]({% link topics/transcriptomics/tutorials/scrna-scanpy-pbmc3k/tutorial.md %}#anndata)).

## CSV to AnnData

> ### {% icon hands_on %} Hands-on: Converting the Data
>
> 1. {% tool [Import Anndata and loom](toolshed.g2.bx.psu.edu/repos/iuc/anndata_import/anndata_import/0.7.5+galaxy0) %} with the following parameters:
>    - *"hd5 format to be created"*: `Anndata file`
>        - *"Format for the annotated data matrix"*: `Tabular, CSV, TSV`
>            - {% icon param-file %} *"Annotated data matrix"*: Multi-select both `SHR` and `WT` datasets
>

> >    > ### {% icon tip %} Tip: Automatic

Reviewer 2: ```suggestion
>    > ### {% icon tip %} Tip: Checking Dimensionality
```

>    >
>    > We can now inspect the dimensionality of the dataset by "peeking" at the dataset in the history and observing the general information.
>    {: .tip}
>    

> >    > ### {% icon comment %} Comment

Reviewer 2: maybe merge into the above tip box? Feels odd that there's two covering the same content

Author: oh whoops I didn't even notice the repeat here

> ### {% icon hands_on %} Hands-on: Filtering
>
> 1. {% tool [Filter](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_filter/scanpy_filter/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Inspect and manipulate** {% icon tool %})
>    - *"Method used for filtering"*: `Filter cell outliers based on counts and numbers of genes expressed, using 'pp.filter_cells'`
>        - *"Filter"*: `Minimum number of genes expressed`
>            - *"Minimum number of genes expressed required for a cell to pass filtering"*: `200`
>
> 1. {% tool [Filter](toolshed.g2.bx.psu.edu/repos/iuc/scanpy_filter/scanpy_filter/1.7.1+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Annotated data matrix"*: `anndata_out` (output of **Filter** {% icon tool %})
>    - *"Method used for filtering"*: `Filter genes based on number of cells or counts, using 'pp.filter_genes'`
>        - *"Filter"*: `Minimum number of cells expressed`
>            - *"Minimum number of cells expressed required for a gene to pass filtering"*: `5`
>

> >    > ### {% icon comment %} Comment
> >    >
> >    > We now set the upper limits to the analysis.
> >    > 
> >    {: .comment}

Reviewer 2: ```suggestion
>    We now set the upper limits to the analysis.
```

## Confounder Removal

> Now we have a set of cells we are reasonably confident contain desirable biological variation (please see the ["An introduction to scRNA-seq data"]({% link videos/#video-transcriptomics-scrna-intro %}) at time point 11:28 as a reference for the types of wanted and unwanted variation). We wish to now regress out the library size variation, as it is not indicative of cell type. To do this we will first ensure that all cells are of the same comparable library size by applying individual size factors to each cell. 

Reviewer 2: ```suggestion
Now we have a set of cells we are reasonably confident contain desirable biological variation (please see the ["An introduction to scRNA-seq data"]({% link videos/watch.md %}?v=transcriptomics/tutorials/scrna-intro/slides&t=11:28) as a reference for the types of wanted and unwanted variation). We wish to now regress out the library size variation, as it is not indicative of cell type. To do this we will first ensure that all cells are of the same comparable library size by applying individual size factors to each cell. 
```

I think this should work

> ### {% icon tip %} Tip: Dimension Reduction
>

> > You can learn more about dimension reduction by consulting the [An introduction to scRNA-seq data analysis]({% link videos/#video-transcriptomics-scrna-intro %}) at time point 13:46.

Reviewer 2: let's see if we can add a syntax for linking to timepoints in the future good to know this is a usecase!

Author: Btw, Is there a good way I can link videos?

Reviewer 2: no :( this is best currently.

I've filed #2465, maybe we can do something better in the near future. There are so many possible ways to include it (e.g. 'mini' player on the tutorial page, etc.) we haven't scratched the surface yet. just need time to do these integrations
